hasher-matcher-actioner/hmalib/lambdas/fetcher.py

```diff
@@ -269,10 +290,35 @@ def _apply_updates_impl(self, delta: tu.ThreatUpdatesDelta) -> None:
                 state.pop(item.key, None)
             else:
                 state[item.key] = item
+                updated[item.key] = item
 
         self._store_state(state.values())
         self._cached_state = state
 
+        post_apply_fn(updated)
+
+    def post_apply(self, updated: t.Dict = {}):
+        """
+        After the fetcher applies an update, check to see if
+        matches for any of the updated signals and if so update
+        their tags.
+        """
+        privacy_group = self.data_s3_key.split("/")[-1].split(".")[0]
+        table = dynamodb.Table(self.data_store_table)
+
+        for update in updated.values():
+            row = update.as_csv_row()
+            # example row format: ('<signal>', '<id>', '<time added>', '<tag1 tags2>')
+            # e.g ('096a6f9...064f', 1234567891234567, '2020-07-31T18:47:45+0000', 'true_positive hma_test')
+            PDQSignalMetadata(
+                signal_id=int(row[1]),
+                ds_id=privacy_group,
+                updated_at=datetime.now(),
+                signal_source="te",
+                signal_hash=row[0],
+                tags=row[3].split(" ") if row[3] else [],
+            ).write_to_table_if_already_exists(table)
```

Comment:
This **feels** like a "patch" update and yet we are doing a PUT.

If there are other systems (there aren't any now), that update the PDQSignalMetadata object in any way, their updates will be lost if the signal were to be refetched.

Can we do this using [UpdateItem](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_UpdateItem.html) instead? And only update the bare minimum? (updated_at, tags)

Reply:
I wrote it this way to avoid the "What is the bare minimum?" question so this could be a DynamoDBItem level method but your point is a good one even if we want to treat the .pdq.te rows to be the ground truth, wiping out the old state when we get an update probably isn't ideal. Will change.